### PR TITLE
feat: support HTTPS for Manager API

### LINKED ISSUE #1706: security: support HTTPS for dashboard
# Feature request

## Please describe your feature

Now dashboard does not support HTTPS access, which is not safe and easy to be hijacked and sniffed.

## Describe the solution you'd like

We should support HTTPS access. How about adding Apache APISIX in front of the manager API?

### PR DESCRIPTION
**Pull request by nic-chen**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

Support for HTTPS by configuring SSL certificate.

**Related issues**

resolve #1706

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

File: api/cmd/managerapi.go [hunk 1]
```diff
@@ -124,6 +124,24 @@ func NewManagerAPICommand() *cobra.Command {
...
+				go func() {
+					err := serverSSL.ListenAndServeTLS(conf.SSLCert, conf.SSLKey)
+					if err != nil && err != http.ErrServerClosed {
+						utils.CloseAll()
+						log.Fatalf("listen and serv fail: %s", err)
+					}
+				}()
+			}
+
 			printInfo()
```
Comment (starsz) [inline api/cmd/managerapi.go:145]: Need to update the printInfo.

File: api/cmd/managerapi.go [hunk 2]
```diff
@@ -124,6 +124,24 @@ func NewManagerAPICommand() *cobra.Command {
...
 
+			// HTTPS
+			if conf.SSLCert != "" && conf.SSLKey != "" {
+				addrSSL := fmt.Sprintf("%s:%d", conf.ServerHost, conf.SSLPort)
+				serverSSL := &http.Server{
+					Addr:         addrSSL,
+					Handler:      r,
+					ReadTimeout:  time.Duration(1000) * time.Millisecond,
+					WriteTimeout: time.Duration(5000) * time.Millisecond,
+				}
```
Comment (bisakhmondal) [inline api/cmd/managerapi.go:135]: Hey, Nice work :+1: . 
I think we should `go` with adding a custom TLS configuration to make it more secure. Also, a go server automatically upgrades requests to HTTP/2 if a tlsconfig is set with HTTPS.

Referring to an awesome blog from [cloudflare](https://blog.cloudflare.com/exposing-go-on-the-internet/).
Thanks :)

File: api/cmd/managerapi.go (hunk 1 above)
Comment (nic-chen) [reply api/cmd/managerapi.go:154]: updated.

File: api/cmd/managerapi.go (hunk 2 above)
Comment (nic-chen) [reply api/cmd/managerapi.go]: updated. thanks.

File: api/cmd/managerapi.go [hunk 3]
```diff
@@ -124,6 +128,29 @@ func NewManagerAPICommand() *cobra.Command {
 				}
 			}()
 
+			// HTTPS
+			if conf.SSLCert != "" && conf.SSLKey != "" {
+				addrSSL := fmt.Sprintf("%s:%d", conf.ServerHost, conf.SSLPort)
```
Comment (tokers) [inline api/cmd/managerapi.go:133]: Use net.JoinHostPort

File: api/conf/conf.yaml [hunk 4]
```diff
@@ -15,12 +15,21 @@
 # limitations under the License.
 #
 
+# yamllint disable rule:comments-indentation
 conf:
-  listen:                 # yamllint disable rule:comments-indentation
```
Comment (gxthrj) [inline api/conf/conf.yaml:19]: If remove this, Will yaml lint report an error?

Comment (nic-chen) [reply api/conf/conf.yaml:19]: @gxthrj  not remove it, just move it to other place.

File: api/cmd/managerapi.go (hunk 3 above)
Comment (nic-chen) [reply api/cmd/managerapi.go]: fixed.

**Comment by imjoey** [comment]
@nic-chen FYI, the CI failure in FE e2e has been fixed in https://github.com/apache/apisix-dashboard/pull/1826. Please rebase the code to get through it.

**Comment by nic-chen** [comment]
> @nic-chen FYI, the CI failure in FE e2e has been fixed in https://github.com/apache/apisix-dashboard/pull/1826. Please rebase the code to get through it.

done, thanks.

**Review by starsz — APPROVED** [review]

File: api/cmd/managerapi.go [hunk 5]
```diff
@@ -124,6 +130,29 @@ func NewManagerAPICommand() *cobra.Command {
...
+						// Causes servers to use Go's default ciphersuite preferences,
+						// which are tuned to avoid attacks. Does nothing on clients.
+						PreferServerCipherSuites: true,
+					},
+				}
+				go func() {
+					err := serverSSL.ListenAndServeTLS(conf.SSLCert, conf.SSLKey)
+					if err != nil && err != http.ErrServerClosed {
+						utils.CloseAll()
+						log.Fatalf("listen and serv fail: %s", err)
```
Comment (tokers) [inline api/cmd/managerapi.go:151]:
```suggestion
						log.Fatalf("listen and serve fail: %s", err)
```

File: api/conf/conf.yaml [hunk 6]
```diff
@@ -15,12 +15,21 @@
...
+  listen:
     # host: 127.0.0.1     # the address on which the `Manager API` should listen.
                           # The default value is 0.0.0.0, if want to specify, please enable it.
                           # This value accepts IPv4, IPv6, and hostname.
     port: 9000            # The port on which the `Manager API` should listen.
+
+  # ssl:
+  #   host: 127.0.0.1     # the address on which the `Manager API` should listen for HTTPS.
+                          # The default value is 0.0.0.0, if want to specify, please enable it.
```
Comment (tokers) [inline api/conf/conf.yaml:28]: We may also support the mTLS.

File: api/cmd/managerapi.go (hunk 5 above)
Comment (nic-chen) [reply api/cmd/managerapi.go]: fixed.

File: api/conf/conf.yaml (hunk 6 above)
Comment (nic-chen) [reply api/conf/conf.yaml:28]: sure. we may support mTLS in the future.

**Review by bisakhmondal — APPROVED** [review]
Perfect :100:

**Review by tokers — APPROVED** [review]

**Review by imjoey — APPROVED** [review]
/lgtm

**Pull request merged**